util/stop: Test flake in TestStopperRunLimitedAsyncTask

This is popping up pretty consistently on my laptop. Here are two example failures:

```
$ make stressrace PKG=./pkg/util/stop/ STRESSFLAGS='-stderr -maxfails 1 -p 100'
go list -tags '' -f 'go test -v -race -tags '\'''\'' -ldflags '\'''\'' -i -c {{.ImportPath}} -o {{.Dir}}/stress.test && (cd {{.Dir}} && if [ -f stress.test ]; then stress -stderr -maxfails 1 -p 100 ./stress.test -test.run '\''.'\'' -test.timeout 10m ; fi)' ./pkg/util/stop/ | /bin/bash
github.com/cockroachdb/cockroach/pkg/util/stop_test
testmain
42 runs so far, 0 failures, over 5s
177 runs so far, 0 failures, over 10s

I161128 10:24:27.050509 11 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.071422 38 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.085797 53 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.098379 39 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.098591 39 util/stop/stopper.go:468  quiescing; tasks left:
1      util/stop/stopper_test.go:147
I161128 10:24:27.107302 22 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.115449 44 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.115520 44 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.115576 44 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.501551 55 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.548743 23 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.577235 82 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.602690 27 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:24:27.602805 27 util/stop/stopper.go:468  quiescing; tasks left:
1      util/stop/stopper_test.go:374
I161128 10:24:28.124249 59 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
--- FAIL: TestStopperRunLimitedAsyncTask (0.51s)
	stopper_test.go:457: expected peak concurrency 4 to equal max concurrency 5
I161128 10:24:28.129965 84 util/stop/stopper.go:294  stopper throttling task from util/stop/stopper_test.go:496 due to semaphore
I161128 10:24:28.130153 83 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
FAIL


ERROR: exit status 1

262 runs completed, 1 failures, over 16s
FAIL
make: *** [stress] Error 1
```

```
$ make stressrace PKG=./pkg/util/stop/ STRESSFLAGS='-stderr -maxfails 1 -p 100'
go list -tags '' -f 'go test -v -race -tags '\'''\'' -ldflags '\'''\'' -i -c {{.ImportPath}} -o {{.Dir}}/stress.test && (cd {{.Dir}} && if [ -f stress.test ]; then stress -stderr -maxfails 1 -p 100 ./stress.test -test.run '\''.'\'' -test.timeout 10m ; fi)' ./pkg/util/stop/ | /bin/bash
github.com/cockroachdb/cockroach/pkg/util/stop_test
testmain

I161128 10:26:45.856864 9 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.306995 10 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.322331 15 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.330647 22 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.330938 22 util/stop/stopper.go:468  quiescing; tasks left:
1      util/stop/stopper_test.go:147
I161128 10:26:46.340925 68 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.350516 27 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.350578 27 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.350621 27 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.364149 82 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.374620 16 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.403395 114 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.412664 41 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
I161128 10:26:46.412783 41 util/stop/stopper.go:468  quiescing; tasks left:
1      util/stop/stopper_test.go:374
I161128 10:26:47.204752 42 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
--- FAIL: TestStopperRunLimitedAsyncTask (0.39s)
	stopper_test.go:457: expected peak concurrency 3 to equal max concurrency 5
I161128 10:26:47.795093 108 util/stop/stopper.go:294  stopper throttling task from util/stop/stopper_test.go:496 due to semaphore
I161128 10:26:47.795196 130 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
FAIL


ERROR: exit status 1

1 runs completed, 1 failures, over 2s
FAIL
make: *** [stress] Error 1
```

--- a/pkg/util/stop/stopper_test.go
+++ b/pkg/util/stop/stopper_test.go
@@ -419,8 +419,9 @@ func TestStopperRunLimitedAsyncTask(t *testing.T) {
 	defer s.Stop()
 
 	const maxConcurrency = 5
-	const duration = 10 * time.Millisecond
+	const numTasks = maxConcurrency * 3
 	sem := make(chan struct{}, maxConcurrency)
+	taskSignal := make(chan struct{}, maxConcurrency)
 	var mu syncutil.Mutex
 	concurrency := 0
 	peakConcurrency := 0
@@ -433,14 +434,30 @@ func TestStopperRunLimitedAsyncTask(t *testing.T) {
 			peakConcurrency = concurrency
 		}
 		mu.Unlock()
-		time.Sleep(duration)
+		<-taskSignal
 		mu.Lock()
 		concurrency--
 		mu.Unlock()
 		wg.Done()
 	}
+	go func() {
+		// Loop until the desired peak concurrency has been reached.
+		for {
+			mu.Lock()
+			c := concurrency
+			mu.Unlock()
+			if c >= maxConcurrency {
+				break
+			}
+			time.Sleep(time.Millisecond)
+		}
+		// Then let the rest of the async tasks finish quickly.
+		for i := 0; i < numTasks; i++ {
+			taskSignal <- struct{}{}
+		}
+	}()
 
-	for i := 0; i < maxConcurrency*3; i++ {
+	for i := 0; i < numTasks; i++ {
 		wg.Add(1)
 		if err := s.RunLimitedAsyncTask(
 			context.TODO(), sem, true /* wait */, f,

